### PR TITLE
refactor: logging technical error separately from violations

### PR DESCRIPTION
**Pull request by perspectivus1**

### CODE REVIEW

File: pkg/validators/readiness/readiness.go [hunk 1]
```diff
@@ -36,6 +39,10 @@ type ReadinessValidator struct {
 	ignoreMissingResources bool
 }
 
+func (v *ReadinessValidator) GetName() string {
+	return ValidatorName
+}
+
 // validates all the resources from readinesslist are ready
 func (v *ReadinessValidator) Validate(ctx context.Context, resources []unstructured.Unstructured) (violations []common.Violation, err error) {
```
Comment (saarzur123) [inline pkg/validators/readiness/readiness.go:47]: Please remove the name from the return variables

File: pkg/validators/readiness/readiness.go [hunk 2]
```diff
@@ -61,38 +68,34 @@ func (v *ReadinessValidator) Validate(ctx context.Context, resources []unstructu
 		if err != nil {
 			msg := fmt.Sprintf("could not determine readiness of resource Kind: %s Name: %s Namespace: %s",
 				resource.GetKind(), resource.GetName(), resource.GetNamespace())
-			v.logger.V(0).Info(msg, "error", err)
+			v.logger.Error(err, msg)
 			continue
```
Comment (saarzur123) [inline pkg/validators/readiness/readiness.go:72]: Please append the errors with `errors.Join()` and return the error

File: pkg/common/types.go [hunk 3]
```diff
@@ -19,11 +20,10 @@ type Validator interface {
 }
```
Comment (saarzur123) [inline pkg/common/types.go:20]: Please remove `ctx context.Context` from `Validate` if you are not using it

File: docs/DETAILS.md [hunk 4]
```diff
@@ -72,10 +72,18 @@ abort:
 Set the value of the `configMapField` to `"true"` to abort the validation. In all other cases (e.g. the `ConfigMap` is not found) the validation will execute normally.
 
 ## Logging
-Kubernetes Resource Validator uses the [`logr`](https://github.com/go-logr/logr) library as an interface to logging violations as well as other errors.
+Kubernetes Resource Validator uses the [`logr`](https://github.com/go-logr/logr) library as its logging interface.
 
 The actual logging implementation is set by users of this library. See the `logr` documentation for a partial [list of supported logging implementations](https://github.com/go-logr/logr#implementations-non-exhaustive).
 
+Note that the `logr` instance should be placed on the `ctx` parameter of the `validation.NewValidation()` function. For example:
```
Comment (saarzur123) [inline docs/DETAILS.md:79]:
```suggestion
# Note that the `logr` instance should be placed on the `ctx` parameter of the `validation.NewValidation()` function. For example:
```

**Review by saarzur123 — CHANGES_REQUESTED** [review]

File: pkg/validation/validation.go [hunk 5]
```diff
@@ -74,7 +74,7 @@ func (v *Validation) preValidate() (aborted bool) {
 			var err error
 			v.Client, err = getClient()
 			if err != nil {
-				v.logger.V(0).Info("unable to create client", "error", err)
+				v.logger.Error(err, "unable to create client", "error", err)
 				panic(err)
```
[the remark this answers is not in the document]
Comment (perspectivus1) [reply pkg/validation/validation.go]: I removed the call to `panic()`.

In my opinion, `shouldAbortValidation()` should never return an error. If no `ConfigMap` that indicates the validation should be aborted is found, we simply continue with the validation...

File: pkg/validation/validation.go [hunk 6]
```diff
@@ -74,7 +74,7 @@ func (v *Validation) preValidate() (aborted bool) {
 			var err error
 			v.Client, err = getClient()
 			if err != nil {
-				v.logger.V(0).Info("unable to create client", "error", err)
+				v.logger.Error(err, "unable to create client", "error", err)
```
[the remark this answers is not in the document]
Comment (perspectivus1) [reply pkg/validation/validation.go]: I removed the second `err` parameter

File: pkg/validation/validation.go [hunk 7]
```diff
@@ -132,12 +136,12 @@ func (v *Validation) readAdditionalResourceTypes(dir string) []schema.GroupVersi
 
 	content, err := afero.ReadFile(v.appFs, additionalResourceTypesFullPath)
 	if err != nil {
-		v.logger.V(2).Info("couldn't find additional resource types file", "file", additionalResourceTypesFullPath)
+		v.logger.V(0).Info("couldn't find additional resource types file", additionalResourceTypesFullPath)
```
[the remark this answers is not in the document]
Comment (perspectivus1) [reply pkg/validation/validation.go:144]: If no `additionalResourceTypes` are found, the flow is still valid. In my opinion this should not result in an error.

File: pkg/validation/validation.go [hunk 8]
```diff
@@ -104,25 +104,29 @@ func (v *Validation) preValidate() (aborted bool) {
...
 	}
 
 	for _, validator := range validators {
 		newViolations, err := validator.Validate(v.ctx, v.Resources)
-		if err == nil && len(newViolations) != 0 {
-			violations = append(violations, newViolations...)
-		}
 		if err != nil {
-			v.logger.V(1).Info("", "error", err)
+			errs = append(errs, err)
```
[the remark this answers is not in the document]
Comment (perspectivus1) [reply pkg/validation/validation.go]: I switched to using `errors.Join()`. Note that the result of a `Join()` is a single error. If you want to deconstruct it to its original errors you would have to do something like this:
```go
errForDeconstruction := cumulativeErr.(interface{ Unwrap() []error })
errs := errForDeconstruction.Unwrap()
```

File: pkg/validation/validation_test.go [hunk 9]
```diff
@@ -77,7 +77,7 @@ var _ = Describe("k8s-resource-validator tests", func() {
 		validation.SetClient(client)
 		Expect(err).To(Succeed())
 
-		violations := validation.Validate(nil)
+		violations, _ := validation.Validate(nil)
```
[the remark this answers is not in the document]
Comment (perspectivus1) [reply pkg/validation/validation_test.go]: Done

File: pkg/validators/readiness/readiness.go (hunk 1 above)
Comment (perspectivus1) [reply pkg/validators/readiness/readiness.go]: I'm not sure what you mean

File: pkg/validation/log.go [hunk 10]
```diff
@@ -9,32 +9,17 @@ import (
...
-This is useful when the logger is remote a service, and minimizing traffic to the logging service is desirable.
+LogViolations() writes violations to a logger.
+It is a convenience method. You may choose to write violations to the log in a different format.
 
 ctx is expected to contain the logger
 
 thresholdLevelForErrors specifies that violations of this level or below are considered errors; others are info
 */
-func Log(ctx context.Context, violations []common.Violation, thresholdLevelForErrors int) {
+func LogViolations(ctx context.Context, violations []common.Violation, thresholdLevelForErrors int) {
```
[the remark this answers is not in the document]
Comment (perspectivus1) [reply pkg/validation/log.go]: Done

File: pkg/validators/readiness/readiness.go (hunk 2 above)
Comment (perspectivus1) [reply pkg/validators/readiness/readiness.go:76]: Done

File: pkg/common/types.go (hunk 3 above)
Comment (perspectivus1) [reply pkg/common/types.go:18]: Done

File: docs/DETAILS.md (hunk 4 above)
Comment (perspectivus1) [reply docs/DETAILS.md]: # I think that formatting the entire line as heading level 1 is a bit much

Instead, I made the **note** bold.

**Comment by perspectivus1** [comment]
@saarzur123 I made a bunch of changes, mostly around error handling: please review.

File: pkg/validation/validation_test.go [hunk 11]
```diff
@@ -235,7 +263,8 @@ var _ = Describe("k8s-resource-validator tests", func() {
 		validation.SetClient(client)
 		Expect(err).To(Succeed())
 
-		grvs := validation.readAdditionalResourceTypes(configDirectory)
+		grvs, err := validation.readAdditionalResourceTypes(configDirectory)
+		Expect(err).To(Not(Succeed()))
```
Comment (saarzur123) [inline pkg/validation/validation_test.go:268]: Please add a case when file not exist, shouldn't return error

**Review by saarzur123 — CHANGES_REQUESTED** [review]

**Comment by perspectivus1** [comment]
@saarzur123, I made all requested changes. Please approve.

**Review by saarzur123 — APPROVED** [review]

**Pull request merged**